### ISSUE TITLE
Review Chia private key & wallet security best practices - implement and document for Machinaris.

### ISSUE THREAD

**Issue by guydavis** [guydavis/machinaris#53]
We inherited the `keys=/root/.chia/mnemonic.txt` behavior from the official Chia Docker, but there needs to be a better way to handle the private key and/or keyring.

Some notes:
- When mode=plotter, we can now just pass farmer_pk and pool_pk, so that's decent.
- But when mode=[harvester|farmer|fullnode], we need a better approach to private key security.

Some references:
- https://thechiaplot.net/secure-farming/

I'm opening this thread and calling for discussion and assistance to find best practices and approaches.  Any contributions of wiki documentation and code pull requests are welcomed.

**Comment by HitLuca** [comment]
Is `mnemonic.txt` needed more than once? Because having it deleted after startup would already be something

**Comment by guydavis** [comment]
Thanks for the suggestion.  Back in late May, when I last tested starting an existing fullnode without invoking `chia keys add`, I was getting errors related to the wallet, perhaps due that Chia version.  

However, based on your suggestion, I have just re-tested this scenario, without mnemonic.txt, and now things are working.  Farming is running and the wallet is available.  Obviously the private keys are still present on disk, but at least the mnemonic.txt is gone.

So I think this is a very good suggestion.  I'll work in the deletion of mnemonic.txt, after verifying the user has it recorded, into a Getting Starting workflow that I'll be adding soon.  Thanks!

**Comment by JakubSido** [comment]
I see more problems... 
If forks, or (sorry) machinaris get access to the phrase ... even if I use a cold wallet, there is still an issue with manipulation with:
1) plot NFT (pool change/self pooling) -- So the attacker can change pool or even change to self pooling 
2) communicate with a pool and change payout address

So, the cold wallet can save my already farmed coins. But the future of the plots and future rewards are compromised. 

Am I wrong? 
Are there any other options to solve those problems?
A cold wallet is not sufficient I think

I have also tried to open this question on official chia github..
See https://github.com/Chia-Network/chia-blockchain/discussions/9007

**Comment by guydavis** [comment]
> I see more problems... If forks, or (sorry) machinaris get access to the phrase ... even if I use a cold wallet, there is still an issue with manipulation with:
> 1. plot NFT (pool change/self pooling) -- So the attacker can change pool or even change to self pooling
> 2. communicate with a pool and change payout address 
> So, the cold wallet can save my already farmed coins. But the future of the plots and future rewards are compromised.

Hi.  Great points! Yes, you are absolutely correct.  The value of plots is tied to the private key encoded in the mnemonic.  For that reason, each person in this blockchain farming ecosystem must assess their own risk/reward position.  For some, that will mean:

1. Only running the [official software](https://github.com/Chia-Network/chia-blockchain), without any 3rd party tools.  They must also ensure their farming systems are not compromised by malware and remote exploits, as we all must do these days using Internet-connected systems.
2. **OR** not even running official farming software, instead using [Flexfarmer](https://www.flexpool.io/get-started/xch/flexfarmer?primaryServer=xch-us-east.flexpool.io) to only harvest for example.
3. **Others** still will want to participate in the blockchain fork community, cross-farming the same plots to obtain those additional rewards.  This, by definition, will expand the exposure surface for one's mnemonic key.  Many will find that an acceptable trade-off, particularly when combined with a cold wallet.

Only each individual can determine their own risk tolerance in pursuit of blockchain rewards by farming plots.  Machinaris will continue to serve those *interested in more than one blockchain* in as secure a manner as possible.  Hope this helps answer your question!

**Comment by JakubSido** [comment]
I am not here to hate :-)
I just really want to find some safe way to do it. 

I believe, there should be another way instead of sharing the keyphrase. 
It should be possible to farm without a wallet sk. 

What about not sharing the phrase but only map private keys needed for farming into the containers and keep unnecessary keys hidden?

**Comment by guydavis** [comment]
Yes, good suggestion.  Like in the official Chia docker, it's possible to set the `keys` environment variable to [persistent](https://github.com/guydavis/machinaris/wiki/Keys#persistent-keys).  Then no mnemonic is required, but all key import and management is responsibility of the advanced user.
